File: chatbot-server/tests/session_client.rs
//! Contract for `static/session-client.js`: owned browser HTTP/session client
//! (MOD009).
//!
//! The shared unit owns fetch bootstrap/CSRF refresh/401 retry, generate
//! retry and voice HTTP helpers with explicit dependencies (fetch, bootstrap
//! fetch, login/CSRF/redirect callbacks, sleep, XHR) and per-instance state
//! (single shared bootstrap promise). `static/chat.js` keeps one instance
//! plus thin adapters and all DOM rendering/callbacks; no duplicate mutable
//! authority. Behavioral coverage lives in
//! `fixtures/session_client_test.js` through the stable import.

use std::path::Path;
use std::process::Command;

fn unit_js() -> &'static str {
    include_str!("../../static/session-client.js")
}

fn chat_js() -> &'static str {
    include_str!("../../static/chat.js")
}

fn function_body<'a>(src: &'a str, fn_name: &str) -> Option<&'a str> {
    let header = format!("function {fn_name}(");
    let start = src.find(&header)?;
    let body = &src[start..];
    let open = body.find('{')?;
    let mut depth = 0i32;
    for (i, ch) in body[open..].char_indices() {
        match ch {
            '{' => depth += 1,
            '}' => {
                depth -= 1;
                if depth == 0 {
                    return Some(&body[open..=open + i]);
                }
            }
            _ => {}
        }
    }
    None
}

fn function_contains(src: &str, fn_name: &str, needle: &str) -> bool {
    function_body(src, fn_name).is_some_and(|body| body.contains(needle))
}

#[test]
fn session_client_projects_owned_retry() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR")).parent().unwrap();
    let run = Command::new("node")
        .arg(root.join("chatbot-server/tests/fixtures/session_client_test.js"))
        .arg(root.join("static/session-client.js"))
        .output()
        .expect("test image must provide the JS behavior-test runtime");
    assert!(
        run.status.success(),
        "JS session-client behavior: {}",
        String::from_utf8_lossy(&run.stderr)
    );
}

#[test]
fn session_client_js_parses() {
    let source = unit_js();
    let allocator = oxc_allocator::Allocator::default();
    let source_type = oxc_span::SourceType::default()
        .with_module(false)
        .with_jsx(false);
    let ret = oxc_parser::Parser::new(&allocator, source, source_type).parse();
    let errors: Vec<String> = ret.diagnostics.errors().map(ToString::to_string).collect();
    assert!(
        errors.is_empty(),
        "static/session-client.js failed to parse ({} error(s)):\n{}",
        errors.len(),
        errors.join("\n")
    );
}

/// Packaging contract: the chat page loads the shared session unit before the
/// application script, and chat.js delegates owned HTTP state to it instead of
/// keeping a second mutable copy.
#[test]
fn chat_page_wires_session_client_before_app_script() {
    let html = include_str!("../../static/templates/chat.html");
    let unit = html
        .find("/static/session-client.js")
        .expect("chat page must load the shared session-client unit");
    let app = html
        .find("/static/chat.js")
        .expect("chat page must load the chat application");
    assert!(
        unit < app,
        "session-client.js must load before chat.js so ChatSessionClient exists"
    );

    let src = chat_js();
    for marker in [
        "ChatSessionClient.createSessionClient",
        "sessionClient.installFetchInterceptor()",
        "sessionClient.refreshSession()",
        "sessionClient.refreshCsrfInit",
        "sessionClient.fetchVoiceRetry",
        "sessionClient.postVoiceSttXhr",
        "sessionClient.withCsrf",
        "sessionClient.handle401OrRetry",
        "sessionClient.fetchWithGenerateRetry",
    ] {
        assert!(
            src.contains(marker),
            "chat.js must delegate owned session HTTP to the shared unit; missing: {marker}"
        );
    }
    for duplicate in [
        "const originalFetch",
        "var sessionRefreshPromise",
        "originalFetch('/login')",
        "xhr.upload.onprogress",
        "new XMLHttpRequest()",
        "name=\"csrf_token\" value",
    ] {
        assert!(
            !src.contains(duplicate),
            "single authority lives in the shared unit; chat.js must not keep {duplicate}"
        );
    }
}

/// Single bootstrap sharing with error recovery: concurrent callers share one
/// attempt, failures clear the promise, guests never bootstrap.
#[test]
fn bootstrap_shares_single_attempt_with_error_recovery() {
    let unit = unit_js();
    assert!(
        unit.contains("var sessionRefreshPromise = null"),
        "one shared bootstrap promise per instance"
    );
    let refresh = function_body(unit, "refreshSession").expect("refreshSession");
    assert!(
        refresh.contains("if (sessionRefreshPromise)")
            && refresh.contains("sessionRefreshPromise.finally")
            && refresh.contains("sessionRefreshPromise = null"),
        "concurrent callers share one attempt and recovery clears it; got: {refresh}"
    );
    assert!(
        refresh.contains("return Promise.resolve(false)")
            && refresh.contains("getLoggedIn()"),
        "guests resolve false without bootstrapping; got: {refresh}"
    );
    assert!(
        refresh.contains("'/login'")
            && refresh.contains("'/login/remember'")
            && refresh.contains("name=\"csrf_token\" value=\"([^\"]+)\"")
            && refresh.contains("'Content-Type': 'application/x-www-form-urlencoded'")
            && refresh.contains("'csrf_token=' + encodeURIComponent"),
        "bootstrap keeps GET /login + form-encoded POST /login/remember; got: {refresh}"
    );
    assert!(
        refresh.contains(".catch(function ()")
            && refresh.contains("return false"),
        "bootstrap failures resolve false for error recovery; got: {refresh}"
    );
    assert!(
        refresh.contains("setCsrfToken(data.csrf_token)"),
        "the restored CSRF is adopted via the explicit setter; got: {refresh}"
    );
    let src = chat_js();
    assert!(
        function_contains(src, "refreshSession", "sessionClient.refreshSession()"),
        "chat keeps only the thin refresh adapter"
    );
}

/// Headers, cookie credentials, and URLs stay exactly as before.
#[test]
fn headers_credentials_urls_preserved() {
    let unit = unit_js();
    let csrf = function_body(unit, "withCsrf").expect("withCsrf");
    assert!(
        csrf.contains("X-CSRF-Token") && csrf.contains("getCsrfToken()"),
        "CSRF header rides the explicit token getter; got: {csrf}"
    );
    let refresh = function_body(unit, "refreshCsrfInit").expect("refreshCsrfInit");
    assert!(
        refresh.contains("x-csrf-token") && refresh.contains("toLowerCase()"),
        "CSRF rebuild stays case-insensitive; got: {refresh}"
    );
    assert!(
        refresh.contains("instanceof Headers") && refresh.contains("forEach"),
        "Headers instances rebuild without mutating the cached init; got: {refresh}"
    );
    assert!(
        !unit.contains("credentials"),
        "cookie credentials stay ambient (same-origin default); no explicit credentials mode"
    );
    assert!(
        unit.contains("'/login'") && unit.contains("'/login/remember'"),
        "bootstrap URLs stay fixed"
    );
    let src = chat_js();
    assert!(
        function_contains(src, "withCsrf", "sessionClient.withCsrf")
            && function_contains(src, "refreshCsrfInit", "sessionClient.refreshCsrfInit"),
        "chat keeps only header thin adapters"
    );
}

/// AbortSignal stays wired through every helper.
#[test]
fn abort_signal_preserved() {
    let unit = unit_js();
    let voice = function_body(unit, "fetchVoiceRetry").expect("fetchVoiceRetry");
    assert!(
        voice.contains("new AbortController()")
            && voice.contains("60000")
            && voice.contains("userAborted")
            && voice.contains("userSignal.aborted")
            && voice.contains("addEventListener('abort'")
            && voice.contains("removeEventListener('abort'"),
        "voice fetch links the user signal to its own 60s stall controller; got: {voice}"
    );
    let stt = function_body(unit, "postVoiceSttXhr").expect("postVoiceSttXhr");
    assert!(
        stt.contains("userSignal.aborted")
            && stt.contains("xhr.abort()")
            && stt.contains("AbortError"),
        "STT XHR aborts on the user signal and surfaces AbortError; got: {stt}"
    );
    let gen = function_body(unit, "fetchWithGenerateRetry").expect("fetchWithGenerateRetry");
    assert!(
        gen.contains("init.signal.aborted") && gen.contains("name = 'AbortError'"),
        "generate retry checks the caller signal before reissuing; got: {gen}"
    );
}

/// Retry eligibility, counts, delays, and per-attempt bodies stay exact.
#[test]
fn retry_eligibility_count_delay_body_preserved() {
    let unit = unit_js();
    let eligible =
        function_body(unit, "isRetryableVoiceStatus").expect("isRetryableVoiceStatus");
    assert!(
        eligible.contains("status === 408")
            && eligible.contains("status === 429")
            && eligible.contains("status === 502")
            && eligible.contains("status >= 500"),
        "voice retry stays 408/429/5xx-only; got: {eligible}"
    );
    let voice = function_body(unit, "fetchVoiceRetry").expect("fetchVoiceRetry");
    assert!(
        voice.contains("attempts = attempts || 3")
            && voice.contains("400 * Math.pow(2, attempts - n)")
            && voice.contains("typeof buildOptions === 'function' ? buildOptions()"),
        "voice fetch retries 3x with exponential backoff and a fresh body per attempt; got: {voice}"
    );
    assert!(
        voice.contains("res.status === 401")
            && voice.contains("Session expired")
            && voice.contains("err.name === 'AbortError' && userAborted"),
        "401 and user aborts never retry; got: {voice}"
    );
    let stt = function_body(unit, "postVoiceSttXhr").expect("postVoiceSttXhr");
    assert!(
        stt.contains("retryableVoice = isRetryableVoiceStatus")
            && stt.contains("retryableVoice = true")
            && stt.contains("retryableVoice === false"),
        "STT marks transport failures retryable and honors the flag; got: {stt}"
    );
    let gen = function_body(unit, "fetchWithGenerateRetry").expect("fetchWithGenerateRetry");
    assert!(
        gen.contains("res.status === 429")
            && gen.contains("res.status === 400 && attempt < 8")
            && gen.contains("attempt < 12")
            && gen.contains("200 + attempt * 150"),
        "generate retries 429 plus early 400s with linear backoff; got: {gen}"
    );
    assert!(
        gen.contains("res.status === 401 && !afterRefresh")
            && gen.contains("refreshSession()")
            && gen.contains("refreshCsrfInit(init)")
            && gen.contains("attempt, true"),
        "generate refreshes once with a rebuilt CSRF header; got: {gen}"
    );
}

/// Guest/native login compatibility: no new auth architecture, no silent race
/// fixes. Guests bypass bootstrap and redirect home; logged-in callers get the
/// response for refresh handling; enc-key 401s never refresh.
#[test]
fn guest_native_login_compatibility_preserved() {
    let unit = unit_js();
    let gate = function_body(unit, "redirectHomeOnAuthFailure")
        .expect("redirectHomeOnAuthFailure");
    assert!(
        gate.contains("getLoggedIn()") && gate.contains("redirectHome()"),
        "the gate reads explicit login state and redirects via callback; got: {gate}"
    );
    let kind = function_body(unit, "response401Kind").expect("response401Kind");
    assert!(
        kind.contains("/encryption key|unlock|invalid encryption key/i"),
        "enc-key 401s classify separately; got: {kind}"
    );
    let handle = function_body(unit, "handle401OrRetry").expect("handle401OrRetry");
    assert!(
        handle.contains("Could not unlock chats")
            && handle.contains("refreshSession()")
            && handle.contains("Session expired. Sign out and log in again."),
        "enc-key throws to unlock, session refreshes once, unrestored throws to sign-in; got: {handle}"
    );
    assert!(
        !unit.contains("/login/keyauth"),
        "the data key must not mint a session"
    );
    let src = chat_js();
    assert!(
        function_contains(src, "redirectHomeOnAuthFailure", "sessionClient.redirectHomeOnAuthFailure")
            && function_contains(src, "handle401OrRetry", "sessionClient.handle401OrRetry"),
        "chat keeps only auth thin adapters"
    );
}

/// Global fetch interception routes through the single allowlist owner.
#[test]
fn global_fetch_interception_semantics_preserved() {
    let unit = unit_js();
    let owner =
        function_body(unit, "isAuthAllowlistedUrl").expect("isAuthAllowlistedUrl");
    for marker in [
        "url.includes('/chat')",
        "url.includes('/regenerate')",
        "url.includes('/get_sets')",
        "url.includes('/load_set')",
        "url.includes('/update_preferences')",
        "url.includes('/history_image')",
    ] {
        assert!(
            owner.contains(marker),
            "allowlist owner must keep {marker}; got: {owner}"
        );
    }
    let install =
        function_body(unit, "installFetchInterceptor").expect("installFetchInterceptor");
    assert!(
        install.contains("isAuthAllowlistedUrl(url)")
            || install.contains("isAuthAllowlistedUrl("),
        "interceptor must route through the single allowlist owner; got: {install}"
    );
    assert!(
        !install.contains("url.includes('/chat')"),
        "allowlist must not duplicate in the interceptor; got: {install}"
    );
    for marker in [
        "instanceof Request",
        "redirectHomeOnAuthFailure()",
        "throw new Error('Session expired')",
    ] {
        assert!(
            install.contains(marker),
            "interceptor must keep {marker}; got: {install}"
        );
    }
    let src = chat_js();
    assert!(
        src.contains("sessionClient.installFetchInterceptor()"),
        "chat installs the owned interceptor instead of wrapping fetch inline"
    );
    assert!(
        !src.contains("window.fetch = function"),
        "single authority lives in the shared unit"
    );
}

/// Callback failures keep precise exception behavior (no broad swallowing).
/// Behavioral regressions live in the fixture; here guards the source shape.
#[test]
fn callback_failures_keep_precise_exceptions() {
    let unit = unit_js();
    let gate = function_body(unit, "redirectHomeOnAuthFailure")
        .expect("redirectHomeOnAuthFailure");
    assert!(
        !gate.contains("try"),
        "redirect failures were uncaught; got: {gate}"
    );
    let refresh = function_body(unit, "refreshSession").expect("refreshSession");
    assert!(
        refresh.contains("setCsrfToken(data.csrf_token)"),
        "setter failure must reach the shared catch (resolve false); got: {refresh}"
    );
    assert!(
        !refresh.contains("try {"),
        "login/token reads must not swallow to defaults; got: {refresh}"
    );
    let csrf = function_body(unit, "withCsrf").expect("withCsrf");
    assert!(
        !csrf.contains("try"),
        "token getter failure must propagate; got: {csrf}"
    );
    let rebuild = function_body(unit, "refreshCsrfInit").expect("refreshCsrfInit");
    assert!(
        !rebuild.contains("try"),
        "token/header reads must propagate; got: {rebuild}"
    );
    let stt = function_body(unit, "postVoiceSttXhr").expect("postVoiceSttXhr");
    assert!(
        stt.contains("userSignal.addEventListener('abort', onUserAbort);"),
        "signal listener failure must reject into the retry loop; got: {stt}"
    );
    assert!(
        !unit.contains("defaultGet")
            && !unit.contains("defaultSetCsrfToken")
            && !unit.contains("defaultRedirectHome")
            && !unit.contains("window.APP_DATA")
            && !unit.contains("window.CSRF_TOKEN")
            && !unit.contains("window.location")
            && !unit.contains("document.querySelector"),
        "no window/document querying in the client; callbacks are explicit required"
    );
    assert!(
        !unit.contains("sleepImpl")
            && !unit.contains("globalFetchNow")
            && !unit.contains("voiceTimeoutMs")
            && !unit.contains("extractFetchUrl")
            && !unit.contains("isRefreshing"),
        "no speculative fallback APIs"
    );
}

File: static/session-client.js
(function (root, factory) {
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = factory();
  } else {
    root.ChatSessionClient = factory();
  }
}(typeof globalThis !== 'undefined' ? globalThis : this, function () {
  'use strict';

  // HTTP/session client boundary: fetch bootstrap, CSRF refresh, 401 retry,
  // generate retry and voice HTTP helpers. One instance owns the single
  // bootstrap promise; chat.js keeps DOM callbacks (location, CSRF meta,
  // login state) and all rendering. No data-key handling here: browsers send
  // the HttpOnly cookie automatically, tests use the header path elsewhere.

  var SESSION_EXPIRED_SEND_MSG =
    'Session expired or unauthorized. Your message was not sent — it is still in the input box.';

  function isAuthAllowlistedUrl(url) {
    if (typeof url !== 'string') return false;
    return url.includes('/chat') ||
      url.includes('/regenerate') ||
      url.includes('/get_sets') ||
      url.includes('/load_set') ||
      url.includes('/create_set') ||
      url.includes('/fork_set') ||
      url.includes('/delete_set') ||
      url.includes('/rename_set') ||
      url.includes('/update_memory') ||
      url.includes('/update_system_prompt') ||
      url.includes('/update_preferences') ||
      url.includes('/delete_message') ||
      url.includes('/reset_chat') ||
      url.includes('/history_pair') ||
      url.includes('/history_image');
  }

  function isRetryableVoiceStatus(status) {
    return status === 408 || status === 429 || status === 502 || status === 503 || status === 504
      || status >= 500;
  }

  function defaultSleep(ms) {
    return new Promise(function (resolve) {
      setTimeout(resolve, ms);
    });
  }

  function defaultXhrFactory() {
    return new XMLHttpRequest();
  }

  // Cohesive owned client. DOM callbacks are explicit and required (no
  // window/document access in this unit; chat.js owns location, CSRF meta,
  // and login state):
  // - fetchImpl: normal requests (defaults to current global fetch at call
  //   time, i.e. the intercepted fetch after installFetchInterceptor).
  // - bootstrapFetch: /login bootstrap (defaults to the captured original
  //   fetch after install, else the normal fetch).
  // - getLoggedIn/getCsrfToken/setCsrfToken/redirectHome: required chat
  //   callbacks; exceptions propagate exactly as the inline code threw.
  // - sleep: backoff waits (tests inject an immediate spy).
  // - xhrFactory: XHR constructor (tests inject a fake).
  // State: one shared bootstrap promise per instance (concurrent callers on
  // the same instance share one attempt; separate instances are independent).
  function createSessionClient(deps) {
    deps = deps || {};
    if (typeof deps.getLoggedIn !== 'function') {
      throw new Error('createSessionClient requires getLoggedIn');
    }
    if (typeof deps.getCsrfToken !== 'function') {
      throw new Error('createSessionClient requires getCsrfToken');
    }
    if (typeof deps.setCsrfToken !== 'function') {
      throw new Error('createSessionClient requires setCsrfToken');
    }
    if (typeof deps.redirectHome !== 'function') {
      throw new Error('createSessionClient requires redirectHome');
    }
    var getLoggedIn = deps.getLoggedIn;
    var getCsrfToken = deps.getCsrfToken;
    var setCsrfToken = deps.setCsrfToken;
    var redirectHome = deps.redirectHome;
    var fetchImpl = (typeof deps.fetchImpl === 'function') ? deps.fetchImpl : null;
    var bootstrapOverride = (typeof deps.bootstrapFetch === 'function') ? deps.bootstrapFetch : null;
    var sleepFn = (typeof deps.sleep === 'function') ? deps.sleep : defaultSleep;
    var xhrFactory = (typeof deps.xhrFactory === 'function') ? deps.xhrFactory : defaultXhrFactory;

    var sessionRefreshPromise = null;
    var installedOriginalFetch = null;

    function getFetch() {
      if (fetchImpl) return fetchImpl;
      return globalThis.fetch;
    }

    function getBootstrapFetch() {
      if (bootstrapOverride) return bootstrapOverride;
      if (installedOriginalFetch) return installedOriginalFetch;
      return getFetch();
    }

    function redirectHomeOnAuthFailure() {
      if (getLoggedIn()) {
        return false;
      }
      redirectHome();
      return true;
    }

    function installFetchInterceptor(target) {
      var globalObj = target || globalThis;
      var originalFetch = globalObj.fetch;
      installedOriginalFetch = originalFetch;
      globalObj.fetch = function (input, init) {
        return originalFetch.apply(this, arguments).then(function (response) {
          if (response.status === 401) {
            var url = input;
            if (typeof Request !== 'undefined' && input instanceof Request) {
              url = input.url;
            }
            if (isAuthAllowlistedUrl(url)) {
              return response;
            }
            if (!redirectHomeOnAuthFailure()) {
              return response;
            }
            throw new Error('Session expired');
          }
          return response;
        });
      };
    }

    // Silent session restore after a server restart: bootstrap a fresh guest
    // session (GET /login) for a CSRF token the dead page cannot produce,
    // exchange the HttpOnly remember cookie (POST /login/remember), then adopt
    // the restored session's CSRF token so same-page requests keep validating.
    // Concurrent callers share one attempt. Resolves to a boolean.
    function refreshSession() {
      if (!getLoggedIn()) {
        return Promise.resolve(false);
      }
      if (sessionRefreshPromise) {
        return sessionRefreshPromise;
      }
      var bootstrap = getBootstrapFetch();
      sessionRefreshPromise = bootstrap('/login')
        .then(function (resp) {
          if (!resp.ok) {
            throw new Error('session bootstrap failed');
          }
          return resp.text();
        })
        .then(function (html) {
          var match = html.match(/name="csrf_token" value="([^"]+)"/);
          if (!match) {
            throw new Error('csrf token not found');
          }
          return bootstrap('/login/remember', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: 'csrf_token=' + encodeURIComponent(match[1]),
          });
        })
        .then(function (resp) {
          if (!resp.ok) {
            return false;
          }
          return resp.json().then(function (data) {
            if (data && data.csrf_token) {
              setCsrfToken(data.csrf_token);
              return true;
            }
            return false;
          });
        })
        .catch(function () {
          return false;
        });
      sessionRefreshPromise.finally(function () { sessionRefreshPromise = null; });
      return sessionRefreshPromise;
    }

    // Rebuild a cached fetch init's X-CSRF-Token header after a session
    // refresh (headers were snapshotted by withCsrf at call time).
    function refreshCsrfInit(init) {
      var token = getCsrfToken();
      if (!init || !init.headers || !token) {
        return init;
      }
      var fresh = Object.assign({}, init);
      var copy;
      if (typeof Headers !== 'undefined' && init.headers instanceof Headers) {
        copy = {};
        init.headers.forEach(function (value, key) { copy[key] = value; });
      } else {
        copy = Object.assign({}, init.headers);
      }
      Object.keys(copy).forEach(function (key) {
        if (key.toLowerCase() === 'x-csrf-token') {
          copy[key] = token;
        }
      });
      fresh.headers = copy;
      return fresh;
    }

    function sleepMs(ms) {
      return sleepFn(ms);
    }

    function fetchVoiceRetry(url, buildOptions, attempts) {
      attempts = attempts || 3;
      function attempt(n) {
        var options = typeof buildOptions === 'function' ? buildOptions() : (buildOptions || {});
        var userSignal = options.signal;
        var controller = new AbortController();
        var timeoutId = setTimeout(function () { controller.abort(); }, 60000);
        var userAborted = false;
        var onUserAbort = function () {
          userAborted = true;
          controller.abort();
        };
        if (userSignal) {
          if (userSignal.aborted) {
            clearTimeout(timeoutId);
            var aborted = new Error('aborted');
            aborted.name = 'AbortError';
            return Promise.reject(aborted);
          }
          userSignal.addEventListener('abort', onUserAbort);
        }
        var opts = Object.assign({}, options, { signal: controller.signal });
        var fetchFn = getFetch();
        return fetchFn(url, opts).then(function (res) {
          if (res.ok) return res;
          if (res.status === 401) {
            redirectHomeOnAuthFailure();
            throw new Error('Session expired');
          }
          if (n > 1 && isRetryableVoiceStatus(res.status)) {
            return sleepMs(400 * Math.pow(2, attempts - n)).then(function () { return attempt(n - 1); });
          }
          throw new Error('request failed (' + res.status + ')');
        }).catch(function (err) {
          if (err && err.message === 'Session expired') throw err;
          if (err && err.name === 'AbortError' && userAborted) throw err;
          if (n <= 1) throw err;
          return sleepMs(400 * Math.pow(2, attempts - n)).then(function () { return attempt(n - 1); });
        }).finally(function () {
          clearTimeout(timeoutId);
          if (userSignal) userSignal.removeEventListener('abort', onUserAbort);
        });
      }
      return attempt(attempts);
    }

    function postVoiceSttXhr(url, buildOptions, attempts) {
      attempts = attempts || 3;
      function attempt(n) {
        var options = typeof buildOptions === 'function' ? buildOptions() : (buildOptions || {});
        var userSignal = options.signal;
        return new Promise(function (resolve, reject) {
          var settled = false;
          function finish(fn, arg) {
            if (!settled) {
              settled = true;
              fn(arg);
            }
          }
          var abortErr = function () {
            var err = new Error('aborted');
            err.name = 'AbortError';
            return err;
          };
          var xhr = xhrFactory();
          xhr.open('POST', url, true);
          var headers = options.headers || {};
          Object.keys(headers).forEach(function (name) {
            xhr.setRequestHeader(name, headers[name]);
          });
          xhr.timeout = 60000;
          var tSend = Date.now();
          var tLastProgress = 0;
          var lastLoaded = 0;
          xhr.upload.onprogress = function (ev) {
            if (!ev || !ev.lengthComputable) return;
            tLastProgress = Date.now();
            lastLoaded = ev.loaded;
          };
          xhr.onload = function () {
            var tDone = Date.now();
            var upMs = tLastProgress > tSend ? tLastProgress - tSend : tDone - tSend;
            var net = {
              bytes: lastLoaded || options.bodyBytes || 0,
              upMs: Math.max(1, Math.round(upMs))
            };
            if (xhr.status === 401) {
              redirectHomeOnAuthFailure();
              finish(reject, new Error('Session expired'));
              return;
            }
            if (xhr.status >= 200 && xhr.status < 300) {
              finish(resolve, { status: xhr.status, responseText: xhr.responseText, net: net });
              return;
            }
            var err = new Error('request failed (' + xhr.status + ')');
            err.retryableVoice = isRetryableVoiceStatus(xhr.status);
            err.net = net;
            finish(reject, err);
          };
          xhr.onerror = function () {
            var err = new Error('network error');
            err.retryableVoice = true;
            finish(reject, err);
          };
          xhr.ontimeout = function () {
            var err = new Error('timeout');
            err.name = 'TimeoutError';
            err.retryableVoice = true;
            finish(reject, err);
          };
          xhr.onabort = function () {
            finish(reject, abortErr());
          };
          var onUserAbort = function () {
            try {
              xhr.abort();
            } catch (_) {
              finish(reject, abortErr());
            }
          };
          if (userSignal) {
            if (userSignal.aborted) {
              finish(reject, abortErr());
              return;
            }
            userSignal.addEventListener('abort', onUserAbort);
          }
          try {
            xhr.send(options.body);
          } catch (sendErr) {
            var serr = new Error(sendErr && sendErr.message ? sendErr.message : 'send failed');
            serr.retryableVoice = true;
            finish(reject, serr);
          }
        }).catch(function (err) {
          if (err && (err.message === 'Session expired' || err.name === 'AbortError')) throw err;
          if (n <= 1 || (err && err.retryableVoice === false)) throw err;
          return sleepMs(400 * Math.pow(2, attempts - n)).then(function () {
            return attempt(n - 1);
          });
        });
      }
      return attempt(attempts);
    }

    function withCsrf(headers) {
      var result = headers ? Object.assign({}, headers) : {};
      var token = getCsrfToken();
      if (token) {
        result['X-CSRF-Token'] = token;
      }
      return result;
    }

    async function withCsrfAsync(headers) {
      var result = headers ? Object.assign({}, headers) : {};
      var token = getCsrfToken();
      if (token) {
        result['X-CSRF-Token'] = token;
      }
      return result;
    }

    async function response401Message(response) {
      try {
        var body = await response.clone().json();
        return (body.error || body.message || '').toString();
      } catch (_) {
        return '';
      }
    }

    async function response401Kind(response) {
      if (response.status !== 401) {
        return null;
      }
      var msg = await response401Message(response);
      if (/encryption key|unlock|invalid encryption key/i.test(msg)) {
        return 'enc_key';
      }
      return 'session';
    }

    async function handle401OrRetry(response, retryFn) {
      var kind = await response401Kind(response);
      if (kind === 'enc_key') {
        throw new Error(
          (await response401Message(response)) ||
            'Could not unlock chats. Sign out and log in with your password.'
        );
      }
      if (response.status === 401) {
        var restored = await refreshSession();
        if (restored && retryFn) {
          return retryFn();
        }
        throw new Error('Session expired. Sign out and log in again.');
      }
      return response;
    }

    function fetchWithGenerateRetry(url, init, attempt, afterRefresh) {
      attempt = attempt || 0;
      var fetchFn = getFetch();
      return fetchFn(url, init).then(function (res) {
        if (res.status === 401 && !afterRefresh) {
          return refreshSession().then(function (restored) {
            if (!restored) {
              redirectHomeOnAuthFailure();
              throw new Error('Session expired');
            }
            return fetchWithGenerateRetry(url, refreshCsrfInit(init), attempt, true);
          });
        }
        if ((res.status === 429 || (res.status === 400 && attempt < 8)) && attempt < 12) {
          return sleepMs(200 + attempt * 150).then(function () {
            if (init && init.signal && init.signal.aborted) {
              var err = new Error('Aborted');
              err.name = 'AbortError';
              throw err;
            }
            return fetchWithGenerateRetry(url, init, attempt + 1, afterRefresh);
          });
        }
        return res;
      });
    }

    return {
      redirectHomeOnAuthFailure: redirectHomeOnAuthFailure,
      installFetchInterceptor: installFetchInterceptor,
      refreshSession: refreshSession,
      refreshCsrfInit: refreshCsrfInit,
      sleepMs: sleepMs,
      isRetryableVoiceStatus: isRetryableVoiceStatus,
      fetchVoiceRetry: fetchVoiceRetry,
      postVoiceSttXhr: postVoiceSttXhr,
      withCsrf: withCsrf,
      withCsrfAsync: withCsrfAsync,
      response401Message: response401Message,
      response401Kind: response401Kind,
      handle401OrRetry: handle401OrRetry,
      fetchWithGenerateRetry: fetchWithGenerateRetry
    };
  }

  return {
    SESSION_EXPIRED_SEND_MSG: SESSION_EXPIRED_SEND_MSG,
    createSessionClient: createSessionClient
  };
}));
